### PR TITLE
RHDEVDOCS-4601-how-to-resize-prometheus-storage-volumes

### PR DESCRIPTION
**Pull request by bburt-rh**
Summary: This PR explains how to resize persistent storage volumes for monitoring components.

- Aligned team: DevTools
- For branches: 4.9+
- Jira: https://issues.redhat.com/browse/RHDEVDOCS-4601
- Direct link to doc preview: https://52277--docspreview.netlify.app/openshift-enterprise/latest/monitoring/configuring-the-monitoring-stack.html#resizing-a-persistent-storage-volume_configuring-the-monitoring-stack
- SME review: @JoaoBraveCoding 
- QE review: @juzhao 
- Peer review: @gabriel-rh

### CODE REVIEW

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:12]:
```suggestion
However, resizing a PV is still possible by using a manual process. If you want to resize a PV for a monitoring component such as Prometheus, Thanos Ruler, or Alertmanager, you can update the appropriate config map where the component is configured, patch the PVC, and then delete and orphan the pods.
```

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:146]: Just would like to mention that users could also resize the size of Alermanager and Prometheus for UWM, not sure if we want to add examples  for that (personally I don't see the need) but just mentioning it

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:171]: I don't think any pods will restart after saving the file, since in this procedure what we will see are new pods spawning once we patch the existing ones
```suggestion
. Save the file to apply the changes.
```

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:173]: Not sure if this warning here makes sense

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:182]: I think it would be better to provide an example that would apply to OCP I have to doublecheck this on the cluster but I think like this it would actually patch the prometheus pvc
```suggestion
# for p in $(oc get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
```

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:178]: See suggestion bellow
```suggestion
. Manually patch every PVC with the updated storage request. The following example resizes the storage size for the `openshift-monitoring` Prometheus component to 100Gi:
```

**Review by JoaoBraveCoding — COMMENTED** [review]
In general it look good, some minor points, I also want to test drive the procedure to be 100% sure it works like we describe it here

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:171]: This is not correct, the pods will indeed re-start

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:182]: I've tested and it works with this label

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (JoaoBraveCoding) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:191]:
```suggestion
# oc delete statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan
```

**Comment by JoaoBraveCoding** [comment]
Tested the procedure with the recommendations done and it worked 👌

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:9]: changed per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:10]: changed per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:12]: changed per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:171]: Ok. Not making any changes here

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:178]: changed per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:182]: changed the label

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:191]: changed

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:146]: I added a note about this at the beginning of this section of the procedure

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:171]: storageClassName: local-storage

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:94]: storageClassName: local-storage

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:23]: add
`A cluster administrator has enabled monitoring for user-defined projects.`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:58]: Specify the size for the storage volume.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:135]: Specify the size for the storage volume.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:60]: that changes the local persistent storage
->
that uses the local persistent storage

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:137]: changes the PVC size to 100 gigabytes
->
sets the PVC size to 100 gigabytes

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:21]: since we are talking about `Resizing a persistent storage volume` here, we'd better mention we've already attached PVs for components,and we want to resize the storage size, so we'd better add one pre-condition:
`You have configured PVCs for core OpenShift Container Platform monitoring components`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:24]: since we are talking about Resizing a persistent storage volume here, so we'd better add one pre-condition:
`You have configured PVCs for components that monitor user-defined projects`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:29]: since we are talking about Resizing a persistent storage volume here
To configure a PVC
change to below is better
`To resize a PVC`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:100]: since we are talking about Resizing a persistent storage volume here
`To configure a PVC`
change to below is better
`To resize a PVC`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:80]: The following example **changes** a PVC that changes local persistent storage for Alertmanager to 40 gigabytes
->
The following example **configures** a PVC that changes local persistent storage for Alertmanager to 40 gigabytes

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:104]: dedicated Alertmanager for user workload monitoring is since 4.11, by default, the dedicated Alertmanager is not enabled , since this doc is for 4.9+, we don't need to mention Alertmanager here

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:137]: The following example changes the PVC size to 100 gigabytes for the Prometheus instance that monitors user-defined projects

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:194]: PVC is restricted to namespace,
```
# for p in $(oc -n openshift-monitoring get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
  oc -n openshift-monitoring patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}'; 
```

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:198]: we could delete this step, reason
1. there is not `--cascade` flag now
```
# oc version
Client Version: 4.12.0-0.nightly-2022-10-05-053337
Kustomize Version: v4.5.4
Server Version: 4.12.0-0.nightly-2022-11-10-033725
Kubernetes Version: v1.25.2+f33d98e

# oc -n openshift-monitoring get statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan
error: unknown flag: --cascade
See 'oc get --help' for usage.
```
2. old pods would be removed automatically and new pods would be restarted automatically

Comment (JoaoBraveCoding) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:198]: I think you typed the wrong command by accident, you used `get` instead of `delete`, I'm not aware of `--cascade` being deprecated

**Review by JoaoBraveCoding — COMMENTED** [review]
After fixing the comments left by @juzhao it's a lgtm from my side

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:94]: fixed

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:21]: Added `You have configured at least one PVC for core {product-title} monitoring components.`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:24]: Added `You have configured at least one PVC for components that monitor user-defined projects.`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:29]: changed

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:100]: changed

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:80]: changed

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:137]: fixed

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:194]: Updated per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:198]: @juzhao @JoaoBraveCoding - Can you please provide guidance on whether or not I need to change anything? Does `--cascade` need to be removed from the command? Thanks!

Comment (juzhao) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:200]: we could remove the step
`Delete the underlying StatefulSet with the --cascade=orphan parameter:`
reason please see my previous comment

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [inline modules/monitoring-resizing-a-persistent-storage-volume.adoc:196]: missed `done` at the end
```
# for p in $(oc -n openshift-monitoring get pvc -l app.kubernetes.io/name=prometheus -o jsonpath='{range .items[*]}{.metadata.name} {end}'); do \
  oc -n openshift-monitoring patch pvc/${p} --patch '{"spec": {"resources": {"requests": {"storage":"100Gi"}}}}';  done
```

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (jan--f) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:200]: I think we do really need this step. This step is specifically called out in the [prometheus-operator docs](https://prometheus-operator.dev/docs/operator/storage/#resizing-volumes) and iiuc is caused by a missing k8s feature.

And @JoaoBraveCoding  is correct, the cascade option should still be there. This is 4.11 but 4.12 should look the same:
```
jan@jflaptop ~/Downloads % oc version
Client Version: 4.11.13
Kustomize Version: v4.5.4
jan@jflaptop ~/Downloads % oc delete --help | grep cascade -C2
	If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.

    --cascade='background':
	Must be "background", "orphan", or "foreground". Selects the deletion cascading strategy for the dependents (e.g. Pods created by a ReplicationController). Defaults to background.
```

**Review by jan--f — COMMENTED** [review]
One missing `done` but otherwise this lgtm.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (juzhao) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:200]: ok, my mistake, I used `oc get` command, which should be `oc delete`, no need to remove, we need this step, lgtm
```
# oc version
Client Version: 4.12.0-0.nightly-2022-10-05-053337
Kustomize Version: v4.5.4
Server Version: 4.12.0-0.nightly-2022-11-10-033725

# oc -n openshift-monitoring delete statefulset -l app.kubernetes.io/name=prometheus --cascade=orphan
statefulset.apps "prometheus-k8s" deleted
```

Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:200]: Thanks, all, for your input and thorough review comments. Very much appreciated.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc]: fixed

**Review by gabriel-rh — COMMENTED** [review]
Looks good! some minor suggestions..

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:9]: I updated both to add `resources` per your suggestion.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:13]: Added in `resource`

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:39]: Leaving as-is

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:96]: The preview hasn't updated, I think. An earlier version incorrectly had asterisks, and I removed them.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:188]: Good suggestion. Made the change. Thanks!

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:12]: I think two sentences is easier to parse. So I've followed your suggestion.

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:205]: good catch. `oc` commands don't require sudo permissions. Fixed.

**Comment by juzhao** [comment]
lgtm, waiting for others to review

**Review by JoaoBraveCoding — APPROVED** [review]
from my side it's lgtm

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:9]: changed per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:139]: changed to `configures` per your suggestion

File: modules/monitoring-resizing-a-persistent-storage-volume.adoc
Comment (bburt-rh) [reply modules/monitoring-resizing-a-persistent-storage-volume.adoc:62]: changed to `configures a PVC that sets the local persistent storage to 100 gigabytes` to avoid repeating the word `configures`

**Comment by bburt-rh** [comment]
@lihongyan1 PTAL at the changes I made per your feedback.

**Pull request merged**
